pkg/backend/backend.go

```diff
@@ -0,0 +1,280 @@
+package backend
+
+import (
+	"context"
+	"time"
+
+	"github.com/bookingcom/carbonapi/pkg/backend/mock"
+	"github.com/bookingcom/carbonapi/pkg/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+)
+
+// The backend to send requests to.
+type Backend struct {
+	BackendImpl
+
+	// Request queues.
+	// There's a separate one for each type, because the stampeding requests of one type should not
+	// impact the other types of requests.
+	// This, for example, removes the dependency between the find and render requests performance.
+	// A big wave of render requests will not have direct impact on the find requests besides through the semphore.
+	renderQ chan *renderReq
+	findQ   chan *findReq
+	infoQ   chan *infoReq // the info requests are expected to be relatively rare
+
+	// The size of the requests semaphore:
+	// The maximum number of simultaneous requests.
+	semaSize int
+
+	requestsInQueue  *prometheus.GaugeVec
+	saturation       prometheus.Gauge
+	timeInQSec       *prometheus.HistogramVec
+	enqueuedRequests *prometheus.CounterVec
+}
+
+// The specific backend implementation.
+//
+// At the moment of writing, it can be one of:
+// * mock, used for tests
+// * net, the standard protocol
+// * grpc, used for streaming gRPC
+type BackendImpl interface {
```

Comment:
`BackendImpl` doesn't sound pretty well IMO as we have `Backend` as well. Maybe we can change it to `BackendQuerier` or `BackendConnector` for example.

Reply:
I think the current name is the better than the suggestions, since it corresponds to a common naming pattern used in multiple projects. It also corresponds to the canonical design pattern and names things for what they are. I'm open to to alternatives, though.